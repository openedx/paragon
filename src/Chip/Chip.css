/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:19 GMT
 */

:root {
  --pgn-spacing-chip-padding-to-icon: 3px;
  --pgn-spacing-chip-padding-icon: .25rem;
  --pgn-spacing-chip-padding-x: .5rem;
  --pgn-spacing-chip-padding-y: .125rem;
  --pgn-spacing-chip-margin: .125rem;
  --pgn-size-chip-icon: 1.25rem;
  --pgn-size-chip-border-radius: .25rem;
}
