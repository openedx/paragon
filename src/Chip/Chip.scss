@import "variables";

@function get-color($element, $color, $type) {
  @return map-get($chip-theme-colors, "#{$color}-#{$element}-#{$type}")
}

@mixin chip-variant($background, $border, $hover-background, $hover-border, $active-background, $active-border, $color, $hover-color, $active-color) {
  background-color: $background;
  color: $color;
  border-color: $border;

  &:hover {
    cursor: pointer;
    text-decoration: none;
    background-color: $hover-background;
    color: $hover-color;
    border-color: $hover-border;
  }

  &.active,
  &:active {
    background-color: $active-background;
    color: $active-color;
    border-color: $active-border;
  }

  &:disabled {
    cursor: default;
    background-color: $background;
    color: $color;
    border-color: $border;
    opacity: $chip-disable-opacity;
    pointer-events: none;
    &::before {
      display: none;
    }
  }
}

.pgn__chip {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: $chip-padding-y $chip-padding-x;
  font-size: $chip-font-size;
  font-weight: $chip-font-weight;
  line-height: $chip-line-height;
  text-align: center;
  white-space: nowrap;
  vertical-align: baseline;
  border-radius: $chip-border-radius;
  border: $chip-border-width solid transparent;
  margin-inline: $chip-margin-inline / 2;

  // Empty chip collapse automatically
  &:empty {
    display: none;
  }

  .chip-icon-before {
    margin-inline-end: 0.5rem;
    margin-left: -0.25em;

    [dir="rtl"] & {
      margin-right: -0.25em;
      margin-left: 0.5rem;
    }
  }

  .chip-icon-after {
    margin-inline-start: 0.5rem;
    margin-right: -0.25em;

    [dir="rtl"] & {
      margin-right: 0.5rem;
      margin-left: -0.25em;
    }
  }
}

@each $color in $chip-theme-variants {
  .pgn__chip-#{$color} {

    $background: get-color("background", $color, "inactive");
    $border: get-color("border", $color, "inactive");
    $text: get-color("text", $color, "inactive");

    $hover-background: get-color("background", $color, "hover");
    $hover-border: get-color("border", $color, "hover");
    $hover-text: get-color("text", $color, "hover");

    $active-background: get-color("background", $color, "selected");
    $active-border: get-color("border", $color, "selected");
    $active-text: get-color("text", $color, "selected");

    @include chip-variant(
      $background,
      $border,
      $hover-background,
      $hover-border,
      $active-background,
      $active-border,
      $text,
      $hover-text,
      $active-text
    );
  }
}