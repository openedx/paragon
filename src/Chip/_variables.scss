$chip-padding-x:                    .5rem !default;
$chip-padding-y:                    .125rem !default;
$chip-padding-to-icon:              3px !default;
$chip-icon-padding:                 .25rem !default;
$chip-margin:                       .125rem !default;
$chip-border-radius:                .25rem !default;
$chip-disable-opacity:              .3 !default;
$chip-icon-size:                    1.25rem !default;

$chip-theme-variants: (
  "light": (
    "background": $light-500,
    "color": $black,
  ),
  "dark": (
    "background": $dark-200,
    "color": $white,
  )
) !default;
