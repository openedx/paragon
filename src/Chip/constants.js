// eslint-disable-next-line import/prefer-default-export
export const STYLE_VARIANTS = {
  DARK: 'dark',
  LIGHT: 'light',
};
