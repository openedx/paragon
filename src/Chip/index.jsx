import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon from '../Icon';

const TAG_VARIANTS = [
  'button',
  'a',
];

const STYLE_VARIANTS = [
  'light',
  'dark',
];

const Chip = ({
  as,
  children,
  className,
  variant,
  active,
  iconBefore,
  iconAfter,
  ...props
}) => {
  const elementType = TAG_VARIANTS.includes(as) ? as : 'button';
  const elementProps = {
    ...props,
    className: classNames(
      'pgn__chip',
      `pgn__chip-${variant}`,
      { active },
      className,
    ),
  };
  const elementContent = (
    <>
      {iconBefore && <Icon className={classNames('chip-icon-before')} src={iconBefore} />}
      {children}
      {iconAfter && <Icon className={classNames('chip-icon-after')} src={iconAfter} />}
    </>
  );
  return React.createElement(props.href ? 'a' : elementType, elementProps, elementContent);
};

Chip.propTypes = {
  /** Specifies the base HTML element. */
  as: PropTypes.elementType,
  /** Specifies an additional `className` to add to the base element. */
  className: PropTypes.string,
  /** Specifies the content of the `Chip`. */
  children: PropTypes.node,
  /** The `Chip` style variant to use. */
  variant: PropTypes.oneOf(STYLE_VARIANTS),
  /** Adds the `active` class to the `Chip`. */
  active: PropTypes.bool,
  /** Disables the `Chip`. */
  disabled: PropTypes.bool,
  /** Providing a `href` will render an `<a>` element, styled as a button. */
  href: PropTypes.string,
  /**
   * An icon component to render before the content.
   * Example import of a Paragon icon component:
   *
   * `import { Check } from '@edx/paragon/icons';`
   */
  iconBefore: PropTypes.node,
  /**
   * An icon component to render before after the content.
   * Example import of a Paragon icon component:
   *
   * `import { Check } from '@edx/paragon/icons';`
   */
  iconAfter: PropTypes.node,
  /** A click handler for the `Chip` */
  onClick: PropTypes.func,
};

Chip.defaultProps = {
  as: 'button',
  className: undefined,
  children: null,
  variant: 'light',
  active: false,
  disabled: false,
  href: undefined,
  iconBefore: undefined,
  iconAfter: undefined,
  onClick: () => {},
};

export default Chip;
