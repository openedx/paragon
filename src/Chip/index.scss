@import "variables";

.pgn__chip {
  background: $light-500;
  border-radius: $chip-border-radius;
  display: inline-flex;
  margin: $chip-margin;
  box-sizing: border-box;

  .pgn__chip__label {
    font-size: $font-size-sm;
    padding: $chip-padding-y $chip-padding-x;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    box-sizing: border-box;
    cursor: default;

    &.p-before {
      padding-left: $chip-padding-to-icon;

      [dir="rtl"] & {
        padding-left: $chip-padding-x;
        padding-right: $chip-padding-to-icon;
      }
    }

    &.p-after {
      padding-right: $chip-padding-to-icon;

      [dir="rtl"] & {
        padding-right: $chip-padding-x;
        padding-left: $chip-padding-to-icon;
      }
    }
  }

  .pgn__chip__icon-before,
  .pgn__chip__icon-after {
    align-items: center;
    display: flex;
    padding-left: $chip-icon-padding;
    padding-right: $chip-icon-padding;
    box-sizing: border-box;
    cursor: default;

    .pgn__icon {
      width: $chip-icon-size;
      height: $chip-icon-size;
    }

    &.active:hover,
    &.active:focus {
      cursor: pointer;
      background: $black;

      * {
        color: $white;
        fill: $white;
      }
    }
  }

  .pgn__chip__icon-before {
    border-radius: $chip-border-radius 0 0 $chip-border-radius;

    [dir="rtl"] & {
      border-radius: 0 $chip-border-radius $chip-border-radius 0;
    }
  }

  .pgn__chip__icon-after {
    border-radius: 0 $chip-border-radius $chip-border-radius 0;

    [dir="rtl"] & {
      border-radius: $chip-border-radius 0 0 $chip-border-radius;
    }
  }

  @each $color, $styles in $chip-theme-variants {
    &.pgn__chip-#{$color} {
      background: map-get($styles, "background");

      * {
        color: map-get($styles, "color");
        fill: map-get($styles, "color");
      }
    }
  }

  &.disabled,
  &:disabled {
    cursor: default;
    opacity: $chip-disable-opacity;
    pointer-events: none;

    &::before {
      display: none;
    }
  }
}
