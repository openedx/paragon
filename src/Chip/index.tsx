import React, { ForwardedRef, KeyboardEventHandler, MouseEventHandler } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
// @ts-ignore
import Icon from '../Icon';

const STYLE_VARIANTS = [
  'light',
  'dark',
];

export interface IChip {
  children: React.ReactNode,
  className?: string,
  variant?: string,
  iconBefore?: React.ReactElement | Function,
  iconAfter?: React.ReactElement | Function,
  onIconBeforeClick?: KeyboardEventHandler & MouseEventHandler,
  onIconAfterClick?: KeyboardEventHandler & MouseEventHandler,
  disabled?: boolean,
}

export const CHIP_PGN_CLASS = 'pgn__chip';

const Chip = React.forwardRef(({
  children,
  className,
  variant,
  iconBefore,
  iconAfter,
  onIconBeforeClick,
  onIconAfterClick,
  disabled,
  ...props
}: IChip, ref: ForwardedRef<HTMLDivElement>) => (
  <div
    className={classNames(
      CHIP_PGN_CLASS,
      `pgn__chip-${variant}`,
      className,
      { disabled },
    )}
    ref={ref}
    {...props}
  >
    {iconBefore && (
      <div
        className={classNames('pgn__chip__icon-before')}
      >
        <Icon src={iconBefore} />
      </div>
    )}
    <div
      className={classNames('pgn__chip__label', {
        'p-before': iconBefore,
        'p-after': iconAfter,
      })}
    >
      {children}
    </div>
    {iconAfter && (
      <div
        className={classNames('pgn__chip__icon-after', { active: onIconAfterClick })}
        role="button"
        onClick={onIconAfterClick}
        onKeyPress={onIconAfterClick}
        tabIndex={disabled ? -1 : 0}
      >
        <Icon src={iconAfter} />
      </div>
    )}
  </div>
));

Chip.propTypes = {
  /** Specifies the content of the `Chip`. */
  children: PropTypes.node.isRequired,
  /** Specifies an additional `className` to add to the base element. */
  className: PropTypes.string,
  /** The `Chip` style variant to use. */
  variant: PropTypes.oneOf(STYLE_VARIANTS),
  /** Disables the `Chip`. */
  disabled: PropTypes.bool,
  /**
   * An icon component to render before the content.
   * Example import of a Paragon icon component:
   *
   * `import { Check } from '@edx/paragon/icons';`
   */
  iconBefore: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  /** A click handler for the `Chip` icon before. */
  onIconBeforeClick: PropTypes.func,
  /**
   * An icon component to render before after the content.
   * Example import of a Paragon icon component:
   *
   * `import { Check } from '@edx/paragon/icons';`
   */
  iconAfter: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  /** A click handler for the `Chip` icon after. */
  onIconAfterClick: PropTypes.func,
};

Chip.defaultProps = {
  className: undefined,
  variant: 'light',
  disabled: false,
  iconBefore: undefined,
  iconAfter: undefined,
  onIconBeforeClick: undefined,
  onIconAfterClick: undefined,
};

export default Chip;
