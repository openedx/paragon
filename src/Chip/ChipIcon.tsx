import React, { KeyboardEventHandler, MouseEventHandler } from 'react';
import PropTypes from 'prop-types';
import Icon from '../Icon';
// @ts-ignore
import IconButton from '../IconButton';
// @ts-ignore
import { STYLE_VARIANTS } from './constants';

export interface ChipIconProps {
  className: string,
  src: React.ReactElement | Function,
  onClick?: KeyboardEventHandler & MouseEventHandler,
  alt?: string,
  variant: string,
  disabled?: boolean,
}

function ChipIcon({
  className, src, onClick, alt, variant, disabled,
}: ChipIconProps) {
  if (onClick) {
    return (
      <IconButton
        className={className}
        src={src}
        onClick={onClick}
        iconAs={Icon}
        alt={alt}
        invertColors={variant === STYLE_VARIANTS.DARK}
        tabIndex={disabled ? -1 : 0}
      />
    );
  }

  return <Icon src={src} className={className} size="sm" />;
}

ChipIcon.propTypes = {
  className: PropTypes.string.isRequired,
  src: PropTypes.oneOfType([PropTypes.element, PropTypes.func]).isRequired,
  onClick: PropTypes.func,
  alt: PropTypes.string,
  variant: PropTypes.string,
  disabled: PropTypes.bool,
};

ChipIcon.defaultProps = {
  onClick: undefined,
  alt: undefined,
  variant: STYLE_VARIANTS.LIGHT,
  disabled: false,
};

export default ChipIcon;
