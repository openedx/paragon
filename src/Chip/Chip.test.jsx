import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { Close } from '../../icons';
import Chip from './index';

function TestChip(props) {
  return (
    <Chip {...props}>
      Test
    </Chip>
  );
}

describe('<Chip />', () => {
  describe('snapshots', () => {
    it('renders without props', () => {
      const tree = renderer.create((
        <TestChip />
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with props iconBefore', () => {
      const tree = renderer.create((
        <TestChip iconBefore={Close} />
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with props iconAfter', () => {
      const tree = renderer.create((
        <TestChip iconAfter={Close} />
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with props iconBefore and iconAfter', () => {
      const tree = renderer.create((
        <TestChip iconBefore={Close} iconAfter={Close}>Chip</TestChip>
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('correct rendering', () => {
    it('renders with correct class when variant is added', () => {
      const wrapper = mount(<TestChip variant="dark" />);
      const chip = wrapper.find('.pgn__chip');
      expect(chip.hasClass('pgn__chip-dark')).toEqual(true);
    });
    it('renders with active class when disabled prop is added', () => {
      const wrapper = mount(<TestChip disabled />);
      const chip = wrapper.find('.pgn__chip');
      expect(chip.hasClass('disabled')).toEqual(true);
    });
    it('renders with the client\'s className', () => {
      const className = 'testClassName';
      const wrapper = mount(<TestChip className={className} />);
      const chip = wrapper.find('.pgn__chip');
      expect(chip.hasClass(className)).toEqual(true);
    });
    it('onIconAfterClick is triggered', () => {
      const func = jest.fn();
      const wrapper = mount(<TestChip iconAfter={Close} onIconAfterClick={func} />);
      const iconAfter = wrapper.find('.pgn__chip__icon-after');
      iconAfter.simulate('click');
      iconAfter.simulate('keypress');
      expect(func).toHaveBeenCalledTimes(2);
    });
  });
});
