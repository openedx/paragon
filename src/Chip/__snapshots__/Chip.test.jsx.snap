// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chip /> snapshots renders with props iconAfter 1`] = `
<div
  className="pgn__chip pgn__chip-light"
>
  <div
    className="pgn__chip__label p-after"
  >
    Test
  </div>
  <div
    className="pgn__chip__icon-after"
    role="button"
    tabIndex={0}
  >
    <span
      className="pgn__icon"
    >
      <svg
        aria-hidden={true}
        fill="none"
        focusable={false}
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Chip /> snapshots renders with props iconBefore 1`] = `
<div
  className="pgn__chip pgn__chip-light"
>
  <div
    className="pgn__chip__icon-before"
  >
    <span
      className="pgn__icon"
    >
      <svg
        aria-hidden={true}
        fill="none"
        focusable={false}
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>
  <div
    className="pgn__chip__label p-before"
  >
    Test
  </div>
</div>
`;

exports[`<Chip /> snapshots renders with props iconBefore and iconAfter 1`] = `
<div
  className="pgn__chip pgn__chip-light"
>
  <div
    className="pgn__chip__icon-before"
  >
    <span
      className="pgn__icon"
    >
      <svg
        aria-hidden={true}
        fill="none"
        focusable={false}
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>
  <div
    className="pgn__chip__label p-before p-after"
  >
    Test
  </div>
  <div
    className="pgn__chip__icon-after"
    role="button"
    tabIndex={0}
  >
    <span
      className="pgn__icon"
    >
      <svg
        aria-hidden={true}
        fill="none"
        focusable={false}
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Chip /> snapshots renders without props 1`] = `
<div
  className="pgn__chip pgn__chip-light"
>
  <div
    className="pgn__chip__label"
  >
    Test
  </div>
</div>
`;
